serial-settings/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+    item: &menu::Item<Context<Settings, Flash>>,
+    args: &[&str],
+    context: &mut Context<Settings, Flash>,
+) {
+    let key = menu::argument_finder(item, args, "item").unwrap().unwrap();
+    let value = menu::argument_finder(item, args, "value").unwrap().unwrap();
+
+    // Now, write the new value into memory.
+    // TODO: Validate it first?
+    match context.settings.set_json(key, value.as_bytes()) {
+        Ok(_) => {
+            context.save();
+            writeln!(
+                context,
+                "Settings in memory may differ from currently operating settings. \
+        Reset device to apply settings."
+            )
+    .unwrap();
+        }
+        Err(e) => {
+            writeln!(context, "Failed to update {key}: {e:?}").unwrap();
+        }
+    }
+}
+
+pub struct SerialTerminal<'a, UsbBus: usb_device::bus::UsbBus, Settings: SerialSettings, Flash: NorFlash + 'static> {
+    usb_device: usb_device::device::UsbDevice<'a, UsbBus>,
+    usb_serial: usbd_serial::SerialPort<'a, UsbBus>,
+    menu: menu::Runner<'a, Context<Settings, Flash>>,
+    output: bbqueue::Consumer<'static, 512>,
+}
+
+impl<'a, UsbBus: usb_device::bus::UsbBus, Settings: SerialSettings, Flash: NorFlash + 'static> SerialTerminal<'a, UsbBus, Settings, Flash> {
+    pub fn new(
+        usb_device: usb_device::device::UsbDevice<'static, UsbBus>,
+        usb_serial: usbd_serial::SerialPort<'static, UsbBus>,
+        input_buffer: &'a mut [u8],
+        settings: Settings,
+        flash: Flash,
+    ) -> Self {
+        let (producer, consumer) = OUTPUT_BUFFER.try_split().unwrap();
+
+        // TODO: Attempt to load from flash first.
+
+        let context = Context {
+            settings,
+            output: OutputBuffer { producer },
+            flash,
+        };
+        Self {
+            menu: menu::Runner::new(make_menu(), input_buffer, context),
+            usb_device,
+            usb_serial,
+            output: consumer,
+        }
+    }
+
+    fn flush(&mut self) {
+        let read = match self.output.read() {
+            Ok(grant) => grant,
+            Err(bbqueue::Error::InsufficientSize) => return,
+            err => err.unwrap(),
+        };
+
+        match self.usb_serial.write(read.buf()) {
+            Ok(count) => read.release(count),
+            Err(usbd_serial::UsbError::WouldBlock) => read.release(0),
+            Err(_) => {
+                let len = read.buf().len();
+                read.release(len);
+            }
+        }
+    }
+
+    pub fn settings(&self) -> &Settings {
+        &self.menu.context.settings
+    }
+
+    pub fn usb_is_suspended(&self) -> bool {
+        self.usb_device.state() == usb_device::device::UsbDeviceState::Suspend
+    }
+
+    pub fn process(&mut self) {
+        self.flush();
+
+        if !self.usb_device.poll(&mut [&mut self.usb_serial]) {
+            return;
+        }
+
+        let mut buffer = [0u8; 64];
+        match self.usb_serial.read(&mut buffer) {
```

Comment:
I think it may be beneficial to pull out any concept of USB and operate on `Read` / `Write` traits instead, as this would allow this to function over i.e. serial UART as well.

Reply:
exactly.